dataproc/job.go

```diff
@@ -0,0 +1,170 @@
+	} `json:"placement"`
+	Reference struct {
+		JobID string `json:"jobId,omitempty"`
+	} `json:"reference,omitempty"`
+	// https://cloud.google.com/dataproc/docs/reference/rest/v1beta2/HadoopJob
+	HadoopJob struct {
+		Args           []string          `json:"args"`
+		MainJarFileURI string            `json:"mainJarFileUri"`
+		FileURIs       []string          `json:"fileUris,omitempty"`
+		Properties     map[string]string `json:"properties,omitempty"`
+	} `json:"hadoopJob"`
+	Status struct {
+		State          string `json:"state,omitempty"`
+		StateStartTime string `json:"stateStartTime,omitempty"`
+		Details        string `json:"details,omitempty"`
+		SubState       string `json:"substate,omitempty"`
+	} `json:"status,omitempty"`
+}
+
+func SubmitJob(j hdfs.Job, svcAcct, project, region, cluster string) error {
+	if j.Mapper == "" || j.Reducer == "" {
+		return errors.New("missing argument Mapper or Reducer")
+	}
+
+	client, err := LoadFromServiceJSON(svcAcct)
+	if err != nil {
+		return err
+	}
+
+	p := make(map[string]string, len(j.Properties))
+	for k, v := range j.Properties {
+		p[k] = v
+	}
+	if _, ok := p["mapred.job.name"]; !ok {
+		p["mapred.job.name"] = j.Name
+	}
+	if _, ok := p["mapred.reduce.tasks"]; !ok {
+		p["mapred.reduce.tasks"] = fmt.Sprintf("%d", j.ReducerTasks)
+	}
+
+	var req jobRequest
+	req.Job.Reference.JobID = j.Name
+	req.Job.Placement.ClusterName = cluster
+	req.Job.HadoopJob.MainJarFileURI = "file:///usr/lib/hadoop-mapreduce/hadoop-streaming.jar"
+	req.Job.HadoopJob.Args = j.JarArgs()
+	req.Job.HadoopJob.FileURIs = j.CacheFiles
+	req.Job.HadoopJob.Properties = p
+
+	resource := fmt.Sprintf("https://dataproc.googleapis.com/v1/projects/%s/regions/%s/jobs:submit", url.PathEscape(project), url.PathEscape(region))
+	job, err := post(client, resource, req)
+	if err != nil {
+		return err
+	}
+	state := job.Status.State
+	log.Printf("job:%s status:%s", job.Reference.JobID, state)
+
+	resource = fmt.Sprintf("https://dataproc.googleapis.com/v1/projects/%s/regions/%s/jobs/%s", url.PathEscape(project), url.PathEscape(region), url.PathEscape(job.Reference.JobID))
+	ticker := time.NewTicker(time.Second * 2)
+	defer ticker.Stop()
+	var i int
+	for range ticker.C {
+		i++
+		job, err = get(client, resource)
+		if err != nil {
+			return err
+		}
+		// if state changes or 30s passes by
+		if state != job.Status.State || i%15 == 0 {
+			state = job.Status.State
+			log.Printf("job:%s status:%s", job.Reference.JobID, state)
+		}
+		if isTerminalState(state) {
+			break
+		}
+	}
+	return nil
+}
+
+func get(client *http.Client, resource string) (*job, error) {
+	resp, err := client.Get(resource)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+	respBody, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+	if resp.StatusCode != 200 {
+		log.Print(string(respBody))
+		return nil, fmt.Errorf("got status code %d", resp.StatusCode)
+	}
+	var j job
+	return &j, json.Unmarshal(respBody, &j)
```

Comment:
nit: i know this works ... but something just feels odd about it ... could pull up the Unmarshal for a bit of clarity